### ISSUE TITLE
Android SDK has incomplete build instructions

### ISSUE THREAD

**Issue by Caerostris** [zerotier/ZeroTierOne#265]
I am unable to compile the ZeroTier Android SDK. The build instructions seem to be fairly incomplete and/or incorrect. The build process failed after trying the following:

Install Android SDK and NDK, as well as ant via brew.
Run `export NDK_BUILD_LOC="/usr/local/Cellar/android-ndk/r10e/ndk-build"`
Run `export ANDROID_PLATFORM="/usr/local/Cellar/android-sdk/24.4.1/platforms/android-21"`
Run `export ZT="~/build/ZeroTierOne"` (this step is missing in the instructions)
Run `ant`  

Result:
`[exec] /Users/caerostris/build/ZeroTierOne/java/obj/local/arm64-v8a/objs/ZeroTierOneJNI//Users/caerostris/build/ZeroTierOne/node/IncomingPacket.o: In function `ZeroTier::IncomingPacket::tryDecode(ZeroTier::RuntimeEnvironment const*, bool)':
     [exec] /Users/caerostris/build/ZeroTierOne/node/IncomingPacket.cpp:61: undefined reference to `ZeroTier::DeferredPackets::enqueue(ZeroTier::IncomingPacket*)'
     [exec] /Users/caerostris/build/ZeroTierOne/java/obj/local/arm64-v8a/objs/ZeroTierOneJNI//Users/caerostris/build/ZeroTierOne/node/Node.o: In function `Node':
     [exec] /Users/caerostris/build/ZeroTierOne/node/Node.cpp:120: undefined reference to `ZeroTier::DeferredPackets::DeferredPackets(ZeroTier::RuntimeEnvironment const*)'
     [exec] /Users/caerostris/build/ZeroTierOne/node/Node.cpp:122: undefined reference to `ZeroTier::DeferredPackets::~DeferredPackets()'
     [exec] /Users/caerostris/build/ZeroTierOne/java/obj/local/arm64-v8a/objs/ZeroTierOneJNI//Users/caerostris/build/ZeroTierOne/node/Node.o: In function `~Node':
     [exec] /Users/caerostris/build/ZeroTierOne/node/Node.cpp:141: undefined reference to `ZeroTier::DeferredPackets::~DeferredPackets()'
     [exec] /Users/caerostris/build/ZeroTierOne/java/obj/local/arm64-v8a/objs/ZeroTierOneJNI//Users/caerostris/build/ZeroTierOne/node/Node.o: In function `ZeroTier::Node::backgroundThreadMain()':
     [exec] /Users/caerostris/build/ZeroTierOne/node/Node.cpp:653: undefined reference to `ZeroTier::DeferredPackets::process()'
     [exec] /Users/caerostris/build/ZeroTierOne/node/Node.cpp:653: undefined reference to `ZeroTier::DeferredPackets::process()'
     [exec] clang++: error: linker command failed with exit code 1 (use -v to see invocation)
     [exec] make: *** [/Users/caerostris/build/ZeroTierOne/java/obj/local/arm64-v8a/libZeroTierOneJNI.so] Error 1

BUILD FAILED
/Users/caerostris/build/ZeroTierOne/java/build.xml:42: exec returned: 2`

System:
OS X El Capitan Version 10.11.1 (15B42)
Java version "1.8.0_51"
Java(TM) SE Runtime Environment (build 1.8.0_51-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.51-b03, mixed mode)
Apache Ant(TM) version 1.9.6 compiled on June 29 2015

**Comment by glimberg** [comment]
It looks like a change to `java/jni/Android.mk` that adds `node/DeferredPackets.cpp` hasn't made it to `main` yet.  The change is present in `dev`.

**Comment by Caerostris** [comment]
Okay, I'll have a look it it! Thanks

**Comment by Caerostris** [comment]
While the `Android.mk` in the `dev` branch works, the CMake file in that branch is referencing files which do not exist, neither in `dev`, nor in `master`. (`../node/Defaults.cpp` and `jni/ZT_jnicache.cpp`).
After removing those two files from the CMake list, I was stuck with an error saying that `jni.h` could not be found.
I found that `CMakeLists.txt` had a static reference to include the Java headers from `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/System/Library/Frameworks/JavaVM.framework/Headers`.
As I'm on 10.11 and not 10.10, the directory obviously wasn't correct.

After changing the version number, I am stuck with
      `[exec]   "ZeroTier::DeferredPackets::enqueue(ZeroTier::IncomingPacket*)", referenced from:`
     `[exec]       ZeroTier::IncomingPacket::tryDecode(ZeroTier::RuntimeEnvironment const*, bool) in IncomingPacket.cpp.o`
     `[exec] ld: symbol(s) not found for architecture i386`
     `[exec] clang: error: linker command failed with exit code 1 (use -v to see invocation)`
     `[exec] make[2]: *** [libZeroTierOneJNI.jnilib] Error 1`
     `[exec] make[1]: *** [CMakeFiles/ZeroTierOneJNI.dir/all] Error 2`
     `[exec] make: *** [all] Error 2`

**Comment by Caerostris** [comment]
I've looked through the commit history of `CMakeLists.txt` and it turned out that while merging two branches in e8cdff3eafd8096da22122eabddf57a09fe2bb90, some changes from 472206dfb23e8c2d285d5cdf19ba1444d07e4d52 have been overwritten with an older version.
`CMakeLists.txt` was also quite incomplete in missing several files, so that compilation worked but linking failed. I have pushed my changes and opened PR #271 for it.
